python/BioSimSpace/FreeEnergy/_ddg.py

```diff
@@ -0,0 +1,495 @@
+        hess=functools.partial(_compute_kappa_hessian, ln_q=ln_q, n=n, factor=factor),
+    )
+
+    if not result.success:
+        raise RuntimeError("The UWHAM minimization failed to converge.")
+
+    f_i = numpy.insert(-result.x, 0, 0.0)
+    ln_z = numpy.insert(result.x, 0, 0.0)
+
+    weights = _compute_weights(ln_z, ln_q, factor)
+
+    if not numpy.allclose(weights.sum(axis=0) / n, 1.0, atol=1e-3):
+        raise RuntimeError("The UWHAM weights do not sum to 1.0")
+
+    df_i = _compute_variance(ln_z, weights, factor, n)
+
+    return f_i, df_i, weights / n
+
+
+def _sort_folders(work_dir):
+    """Sorts folder names by lambda value, ensuring they are read correctly.
+
+    Parameters
+    ----------
+    work_dir : str
+        The directory containing the simulation data.
+
+    Returns
+    -------
+    folders : dict
+        A dictionary of folder names and their corresponding lambda values.
+    """
+    folders = {}
+    for folder in pathlib.Path(work_dir).iterdir():
+        if folder.is_dir() and folder.name.startswith("lambda_"):
+            try:
+                lambda_val = float(folder.name.split("_")[-1])
+            except ValueError:
+                continue
+            folders[lambda_val] = folder
+    return {k: v for k, v in sorted(folders.items())}
+
+
+def _get_inflection_indices(folders):
+    # Find folders at which 'direction' goes from 1 to -1
+    # This is the point at which the direction of the lambda windows changes
+    # NOTE: this assumes that the folders are correctly sorted
+
+    # check that the keys are sorted
+    keys = list(folders.keys())
+    if keys != sorted(keys):
+        raise ValueError(f"Folders are not sorted correctly. {keys} != {sorted(keys)}")
+
+    directions = []
+    for folder in folders.values():
+        df = _pd.read_csv(folder / "openmm.csv")
+        direction = df["direction"].values[0]
+        directions.append(direction)
+
+    # get the indices at which the direction changes
+    for i in range(len(directions) - 1):
+        if directions[i] != directions[i + 1]:
+            inflection_indices = (i, i + 1)
+            break
+
+    return inflection_indices
+
+
+def analyse_UWHAM(work_dir, ignore_lower, ignore_upper, inflection_indices=None):
+    """
+    Analyse the output of BioSimSpace AToM simulations.
+
+    Parameters
+    ----------
+    work_dir : str
+        The directory containing the simulation data.
+    ignore_lower : int
+        The number of rows to ignore at the start of each file.
+    ignore_upper : int
+        The number of rows to ignore at the end of each file.
+    inflection_indices : tuple, optional
+        The point at which 'direction' changes.
+        Should be (last index of direction 1, first index of direction 2).
+        If not provided not provided, will be implied from files.
+
+    Returns
+    -------
+    ddg_total : float
+        The free energy.
+    ddg_total_error : float
+        The error in the free energy.
```

Comment:
Can these be `BioSimSpace.Types.Energy` types.

Reply:
ddg now returns energies